src/Disks/IO/CachedOnDiskReadBufferFromFile.cpp

```diff
@@ -840,6 +848,11 @@ bool CachedOnDiskReadBufferFromFile::nextImplStep()
 
     chassert(!internal_buffer.empty());
 
+    /// We allocate buffers not less than 1M so that s3 requests will not be too small. But the same buffers (members of AsynchronousReadIndirectBufferFromRemoteFS)
+    /// are used for reading from files. Some of these readings are fairly small and their performance degrade when we use big buffers (up to ~20% for queries like Q23 from ClickBench).
+    if (use_external_buffer && read_type == ReadType::CACHED && settings.local_fs_buffer_size < internal_buffer.size())
+        internal_buffer.resize(settings.local_fs_buffer_size);
```

Comment:
below when calling predownload() we might decide to bypass cache and change read_method, so we need to resize the buffer back somewhere there

Reply:
if I'm not missing smth, we do resize only if `read_type == ReadType::CACHED` and in the only place where we assign positive value to `bytes_to_predownload` we also set `read_type = REMOTE_FS_READ_AND_PUT_IN_CACHE`

Reply:
ah, yes, sorry